File: src/test/java/me/pcasaes/hexoids/EnforceCleanArchitectureDependencyFlowArchUnitTest.java
package me.pcasaes.hexoids;

import com.tngtech.archunit.core.domain.JavaClasses;
import com.tngtech.archunit.core.importer.ClassFileImporter;
import com.tngtech.archunit.lang.ArchRule;
import org.junit.jupiter.api.Test;

import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;

public class EnforceCleanArchitectureDependencyFlowArchUnitTest {

    @Test
    void testDomainsShouldNotDependOnOtherLayers() {
        JavaClasses importedClasses = new ClassFileImporter()
                .importPackages("me.pcasaes.hexoids.core.domain..");

        ArchRule rule = noClasses()
                .should()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.infrastructure..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.entrypoints..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.core.application..");


        rule.check(importedClasses);
    }

    @Test
    void testApplicationShouldOnlyDependOnDomain() {
        JavaClasses importedClasses = new ClassFileImporter()
                .importPackages("me.pcasaes.hexoids.core.application..");

        ArchRule rule = noClasses()
                .should()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.infrastructure..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.entrypoints..");


        rule.check(importedClasses);
    }

    @Test
    void testEntrypointShouldOnlyDependOnApplicationAndDomain() {
        JavaClasses importedClasses = new ClassFileImporter()
                .importPackages("me.pcasaes.hexoids.entrypoints..");

        ArchRule rule = noClasses()
                .should()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.infrastructure..");


        rule.check(importedClasses);
    }

    @Test
    void testInfrastructureShouldOnlyDependOnApplicationAndDomain() {
        JavaClasses importedClasses = new ClassFileImporter()
                .importPackages("me.pcasaes.hexoids.infrastructure..");

        ArchRule rule = noClasses()
                .should()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
                .orShould()
                .dependOnClassesThat()
                .resideInAnyPackage("me.pcasaes.hexoids.entrypoints..");


        rule.check(importedClasses);
    }
}

File: src/test/kotlin/me/pcasaes/hexoids/EnforceCleanArchitectureDependencyFlowArchUnitTest.kt
package me.pcasaes.hexoids

import com.tngtech.archunit.core.importer.ClassFileImporter
import com.tngtech.archunit.lang.ArchRule
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition
import org.junit.jupiter.api.Test

class EnforceCleanArchitectureDependencyFlowArchUnitTest {
    @Test
    fun testDomainsShouldNotDependOnOtherLayers() {
        val importedClasses = ClassFileImporter()
            .importPackages("me.pcasaes.hexoids.core.domain..")

        val rule: ArchRule = ArchRuleDefinition.noClasses()
            .should()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.infrastructure..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.entrypoints..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.core.application..")


        rule.check(importedClasses)
    }

    @Test
    fun testApplicationShouldOnlyDependOnDomain() {
        val importedClasses = ClassFileImporter()
            .importPackages("me.pcasaes.hexoids.core.application..")

        val rule: ArchRule = ArchRuleDefinition.noClasses()
            .should()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.infrastructure..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.entrypoints..")


        rule.check(importedClasses)
    }

    @Test
    fun testEntrypointShouldOnlyDependOnApplicationAndDomain() {
        val importedClasses = ClassFileImporter()
            .importPackages("me.pcasaes.hexoids.entrypoints..")

        val rule: ArchRule = ArchRuleDefinition.noClasses()
            .should()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.infrastructure..")


        rule.check(importedClasses)
    }

    @Test
    fun testInfrastructureShouldOnlyDependOnApplicationAndDomain() {
        val importedClasses = ClassFileImporter()
            .importPackages("me.pcasaes.hexoids.infrastructure..")

        val rule: ArchRule = ArchRuleDefinition.noClasses()
            .should()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.configuration..")
            .orShould()
            .dependOnClassesThat()
            .resideInAnyPackage("me.pcasaes.hexoids.entrypoints..")


        rule.check(importedClasses)
    }
}
